// @cerebral/http/lib/operators/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _httpDelete = require('./httpDelete');

Object.defineProperty(exports, 'httpDelete', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpDelete).default;
  }
});

var _httpGet = require('./httpGet');

Object.defineProperty(exports, 'httpGet', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpGet).default;
  }
});

var _httpPatch = require('./httpPatch');

Object.defineProperty(exports, 'httpPatch', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpPatch).default;
  }
});

var _httpPost = require('./httpPost');

Object.defineProperty(exports, 'httpPost', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpPost).default;
  }
});

var _httpPut = require('./httpPut');

Object.defineProperty(exports, 'httpPut', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpPut).default;
  }
});

var _httpUploadFile = require('./httpUploadFile');

Object.defineProperty(exports, 'httpUploadFile', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_httpUploadFile).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
