// cerebral/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parallel = exports.sequence = exports.provide = exports.compute = undefined;
exports.Controller = Controller;
exports.UniversalController = UniversalController;

var _Compute = require('./Compute');

Object.defineProperty(exports, 'compute', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_Compute).default;
  }
});

var _Provide = require('./Provide');

Object.defineProperty(exports, 'provide', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_Provide).default;
  }
});

var _functionTree = require('function-tree');

Object.defineProperty(exports, 'sequence', {
  enumerable: true,
  get: function get() {
    return _functionTree.sequence;
  }
});
Object.defineProperty(exports, 'parallel', {
  enumerable: true,
  get: function get() {
    return _functionTree.parallel;
  }
});

var _Controller = require('./Controller');

var _Controller2 = _interopRequireDefault(_Controller);

var _UniversalController = require('./UniversalController');

var _UniversalController2 = _interopRequireDefault(_UniversalController);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Controller(options) {
  return new _Controller2.default(options);
}
function UniversalController(options) {
  return new _UniversalController2.default(options);
}
